// REFERENCE: RuleMethod-exception-syntax
using Nitra;
using System.Console;

module Program
{
  Main() : void
  {
    def source = SourceSnapshot("test");
    def parserHost = ParserHost();
    def result = ExceptionTest1.Start(source, parserHost);
    if (result.IsSuccess)
    {
      def ast = ExceptionTest1Ast.Start.Create(result);
      try
      {
        _ = ast.Foo(-1);
        WriteLine("Exception not thrown");
      }
      catch
      {
        | e => WriteLine(e.Message);
      }

      // verify that recursion detection flag is cleared after exception propagation
      WriteLine(ast.Foo(1));
    }
    else
      WriteLine("Parse failed");
  }
}
/*
BEGIN-OUTPUT
Value should be greater than zero.
2
END-OUTPUT
*/
